pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/refreshsegment/RefreshSegmentTaskGenerator.java

```diff
@@ -54,84 +54,99 @@ public String getTaskType() {
   public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
-    String taskType = RefreshSegmentTask.TASK_TYPE;
     List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
-    PinotHelixResourceManager pinotHelixResourceManager = _clusterInfoAccessor.getPinotHelixResourceManager();
-
-    int tableNumTasks = 0;
-
     for (TableConfig tableConfig : tableConfigs) {
-      String tableNameWithType = tableConfig.getTableName();
-      LOGGER.info("Start generating RefreshSegment tasks for table: {}", tableNameWithType);
-
       // Get the task configs for the table. This is used to restrict the maximum number of allowed tasks per table at
       // any given point.
       Map<String, String> taskConfigs;
       TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
       if (tableTaskConfig == null) {
-        LOGGER.warn("Failed to find task config for table: {}", tableNameWithType);
+        LOGGER.warn("Failed to find task config for table: {}", tableConfig.getTableName());
         continue;
       }
       taskConfigs = tableTaskConfig.getConfigsForTaskType(RefreshSegmentTask.TASK_TYPE);
-      Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: %s", tableNameWithType);
-      int tableMaxNumTasks = RefreshSegmentTask.MAX_NUM_TASKS_PER_TABLE;
-      String tableMaxNumTasksConfig = taskConfigs.get(MinionConstants.TABLE_MAX_NUM_TASKS_KEY);
-      if (tableMaxNumTasksConfig != null) {
-        try {
-          tableMaxNumTasks = Integer.parseInt(tableMaxNumTasksConfig);
-        } catch (Exception e) {
-          tableMaxNumTasks = RefreshSegmentTask.MAX_NUM_TASKS_PER_TABLE;
-          LOGGER.warn("MaxNumTasks have been wrongly set for table : {}, and task {}", tableNameWithType, taskType);
-        }
+      Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: %s",
+          tableConfig.getTableName());
+
+      pinotTaskConfigs.addAll(generateTasksForTable(tableConfig, taskConfigs));
+    }
+
+    return pinotTaskConfigs;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(TableConfig tableConfig, Map<String, String> taskConfigs)
+      throws Exception {
+    Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: %s", tableConfig.getTableName());
+    return generateTasksForTable(tableConfig, taskConfigs);
+  }
+
+  private List<PinotTaskConfig> generateTasksForTable(TableConfig tableConfig, Map<String, String> taskConfigs) {
+    String taskType = RefreshSegmentTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+    PinotHelixResourceManager pinotHelixResourceManager = _clusterInfoAccessor.getPinotHelixResourceManager();
+
+    String tableNameWithType = tableConfig.getTableName();
+    LOGGER.info("Start generating RefreshSegment tasks for table: {}", tableNameWithType);
+
+
+    int tableNumTasks = 0;
+    int tableMaxNumTasks = RefreshSegmentTask.MAX_NUM_TASKS_PER_TABLE;
```

Comment:
not introduced in this PR but any rationale for keeping RefreshSegmentTask.MAX_NUM_TASKS_PER_TABLE default value as 20.

Reply:
No particular reason - chose a low value to make sure we don't put stress on deployments that have few minions. This can be overriden with table config.

Reply:
Hmm one improvement can be to keep it equal to number of partitions in a table and process one segment per partition (or instance) by default. This way the load gets normalized across hosts. We can do this as a follow-up too.

Reply:
Good point. Can you please add these thoughts to this issue - https://github.com/apache/pinot/issues/14483 where we can do it as a followup?